qurro/tests/test_qarcoal.py

```diff
@@ -155,24 +220,223 @@ def test_samples_to_use(self, get_mp_data):
-    feats = ["F{}".format(i) for i in range(6)]
-    table = biom.table.Table(mat, feats, samps)
-    tax_labels = [
-        "Bulbasaur",
-        "Ivysaur",
-        "Venusaur",
-        "Charmander",
-        "Charmeleon",
-        "Charizard",
-    ]
-    confidence = ["0.99"] * 6
-    taxonomy = pd.DataFrame([feats, tax_labels, confidence]).T
-    taxonomy.columns = ["feature-id", "Taxon", "Confidence"]
-    taxonomy.set_index("feature-id", inplace=True, drop=True)
-    qarcoal(table, Metadata(taxonomy), "Char", "saur")
+class TestIrregularData():
+    @pytest.fixture(scope="class")
+    def get_testing_data(self):
+        mat = [np.random.randint(1, 10) for x in range(30)]
+        mat = np.reshape(mat, (6, 5))
+        samps = ["S{}".format(i) for i in range(5)]
+        feats = ["F{}".format(i) for i in range(6)]
+        table = biom.table.Table(mat, feats, samps)
+        tax_labels = [
+            "Bulbasaur",
+            "Ivysaur",
+            "Venusaur",
+            "Charmander",
+            "Charmeleon",
+            "Charizard",
+        ]
+        confidence = ["0.99"] * 6
+        taxonomy = pd.DataFrame([
+            feats,
+            tax_labels,
+            confidence,
+        ]).T
+        taxonomy.columns = ["feature-id", "Taxon", "Confidence"]
+        taxonomy.set_index("feature-id", inplace=True, drop=True)
+        data = namedtuple("Data", "table taxonomy")
+        return data(table, taxonomy)
+
+    def test_taxonomy_missing_features(self, get_testing_data):
+        """Taxonomy file missing features that are present in feature table"""
+        table = get_testing_data.table.to_dataframe()
+        taxonomy = get_testing_data.taxonomy
+        taxonomy_miss = taxonomy.iloc[:-1, :]  # remove F5 from taxonomy
+        num_df, denom_df = filter_and_join_taxonomy(
+            get_testing_data.table.to_dataframe(),
+            taxonomy_miss,
+            "Char",
+            "saur",
+        )
+
+        # F5 dropped out
+        assert "F5" not in num_df.index
+
+        num_features = ["F3", "F4"]
+        denom_features = ["F1", "F2"]
+        sample_order = ["S{}".format(i) for i in range(5)]
+
+        table_num_taxon_filt = pd.DataFrame(table.loc[num_features])
+        table_denom_taxon_filt = pd.DataFrame(table.loc[denom_features])
+
+        num_df_taxon_filt = num_df.loc[num_features]
+        denom_df_taxon_filt = denom_df.loc[denom_features]
+
+        # set operations in Qarcoal change order of samples
+        table_num_taxon_filt = table_num_taxon_filt[sample_order]
+        table_denom_taxon_filt = table_denom_taxon_filt[sample_order]
+        num_df_taxon_filt = num_df_taxon_filt[sample_order]
+        denom_df_taxon_filt = denom_df_taxon_filt[sample_order]
+
+        # test that num/denom df accurately extract values from table
+        assert table_num_taxon_filt.equals(num_df_taxon_filt)
+        assert table_denom_taxon_filt.equals(denom_df_taxon_filt)
```

Comment:
It looks like the ~9 lines above here (starting with `# set operations in Qarcoal...`) are repeated throughout a few of these tests. Do you think it'd be reasonable to split this off into its own utility function for the tests?

The main reason for this IMO is that it's kind of hard to tell at first glance what each of these tests are doing, so removing redundant code should make these easier to come back to in the future.